lib/value/src/kind/accessor.rs

```diff
@@ -0,0 +1,543 @@
+                        }
+
+                        // Re-insert all elements with the correct index applied.
+                        for (i, kind) in entries {
+                            collection.known_mut().insert((i - 1).into(), kind);
+                        }
+
+                        Some(kind)
+                    })?;
+
+                    return Some(kind);
+                }
+
+                // Try finding the index in the existing array.
+                Segment::Index(index) => usize::try_from(*index).ok().and_then(|index| {
+                    kind.array
+                        .as_mut()
+                        .and_then(|collection| collection.known_mut().get_mut(&(index.into())))
+                })?,
+            };
+        }
+
+        None
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use std::collections::HashMap;
+
+    use lookup::LookupBuf;
+
+    use super::*;
+
+    #[test]
+    #[allow(clippy::too_many_lines)]
+    fn test_find_at_path() {
+        struct TestCase {
+            kind: Kind,
+            path: LookupBuf,
+            want: Option<Kind>,
+        }
+
+        for (title, TestCase { kind, path, want }) in HashMap::from([
+            (
+                "primitive",
+                TestCase {
+                    kind: Kind::bytes(),
+                    path: "foo".into(),
+                    want: None,
+                },
+            ),
+            (
+                "multiple primitives",
+                TestCase {
+                    kind: Kind::integer().or_regex(),
+                    path: "foo".into(),
+                    want: None,
+                },
+            ),
+            (
+                "object w/ matching path",
+                TestCase {
+                    kind: Kind::object(BTreeMap::from([("foo".into(), Kind::integer())])),
+                    path: "foo".into(),
+                    want: Some(Kind::integer()),
+                },
+            ),
+            (
+                "object w/o matching path",
+                TestCase {
+                    kind: Kind::object(BTreeMap::from([("foo".into(), Kind::integer())])),
+                    path: "bar".into(),
+                    want: None,
+                },
+            ),
+            (
+                "array w/ matching path",
+                TestCase {
+                    kind: Kind::array(BTreeMap::from([(1.into(), Kind::integer())])),
+                    path: LookupBuf::from_str("[1]").unwrap(),
+                    want: Some(Kind::integer()),
+                },
+            ),
+            (
+                "array w/o matching path",
+                TestCase {
+                    kind: Kind::array(BTreeMap::from([(1.into(), Kind::integer())])),
+                    path: LookupBuf::from_str("[2]").unwrap(),
+                    want: None,
+                },
+            ),
+            (
+                "array w/ matching path, shifting indices",
```

Comment:
I suspect I copied this example from the `remote_at_path` tests, since they are common in structure, just with different results. This one doesn't make sense for `find_at_path` though, since no mutation happens of the original event, will update 👍